src/rigid.rs

```diff
@@ -0,0 +1,265 @@
+use approxeq::ApproxEq;
+use num_traits::Float;
+use trig::Trig;
+use {TypedRotation3D, TypedTransform3D, TypedVector3D, UnknownUnit};
+
+/// A rigid transformation. All lengths are preserved under such a transformation.
+///
+///
+/// Internally, this is a rotation and a translation, with the rotation
+/// applied first (i.e. `Rotation * Translation`, in row-vector notation)
+///
+/// This can be more efficient to use over full matrices, especially if you
+/// have to deal with the decomposed quantities often.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[repr(C)]
+pub struct TypedRigidTransform3D<T, Src, Dst> {
+    pub rotation: TypedRotation3D<T, Src, Dst>,
+    pub translation: TypedVector3D<T, Dst>,
+}
+
+pub type RigidTransform3D<T> = TypedRigidTransform3D<T, UnknownUnit, UnknownUnit>;
+
+// All matrix multiplication in this file is in row-vector notation,
+// i.e. a vector `v` is transformed with `v * T`, and if you want to apply `T1`
+// before `T2` you use `T1 * T2`
+
+impl<T: Float + ApproxEq<T>, Src, Dst> TypedRigidTransform3D<T, Src, Dst> {
+    /// Construct a new rigid transformation, where the `rotation` applies first
+    #[inline]
+    pub fn new(rotation: TypedRotation3D<T, Src, Dst>, translation: TypedVector3D<T, Dst>) -> Self {
+        Self {
+            rotation,
+            translation,
+        }
+    }
+
+    /// Construct an identity transform
+    #[inline]
+    pub fn identity() -> Self {
+        Self {
+            rotation: TypedRotation3D::identity(),
+            translation: TypedVector3D::zero(),
+        }
+    }
+
+    /// Construct a new rigid transformation, where the `translation` applies first
+    #[inline]
+    pub fn new_from_reversed(
+        translation: TypedVector3D<T, Src>,
+        rotation: TypedRotation3D<T, Src, Dst>,
+    ) -> Self {
+        // T * R
+        //   = (R * R^-1) * T * R
+        //   = R * (R^-1 * T * R)
+        //   = R * T'
+        //
+        // T' = (R^-1 * T * R) is also a translation matrix
+        // It is equivalent to the translation matrix obtained by rotating the
+        // translation by R
+
+        let translation = rotation.rotate_vector3d(&translation);
+        Self {
+            rotation,
+            translation,
+        }
+    }
+
+    #[inline]
+    pub fn from_rotation(rotation: TypedRotation3D<T, Src, Dst>) -> Self {
+        Self {
+            rotation,
+            translation: TypedVector3D::zero(),
+        }
+    }
+
+    #[inline]
+    pub fn from_translation(translation: TypedVector3D<T, Dst>) -> Self {
+        Self {
+            translation,
+            rotation: TypedRotation3D::identity(),
+        }
+    }
+
+    /// Decompose this into a position and an orientation to be applied in the opposite order
```

Comment:
Change "position" to "translation" and "orientation" to "rotation" for consistency?